File: src/project/config-parser.ts
import fs from 'node:fs';
import path from 'node:path';
import YAML from 'yaml';
import { VideoCutError } from '../errors.js';

/** 项目配置结构 */
export interface ProjectConfig {
  template: string;
  preset?: string;
  quality?: string;
  output?: {
    filename?: string;
  };
  variables: Record<string, unknown>;
}

/**
 * 解析项目配置文件（YAML 或 JSON）
 */
export function parseProjectConfig(configPath: string): ProjectConfig {
  const absPath = path.resolve(configPath);

  if (!fs.existsSync(absPath)) {
    throw new VideoCutError(`配置文件不存在: ${absPath}`);
  }

  const raw = fs.readFileSync(absPath, 'utf-8');
  const ext = path.extname(absPath).toLowerCase();

  let parsed: unknown;
  try {
    if (ext === '.json') {
      parsed = JSON.parse(raw);
    } else {
      // .yaml, .yml or other — treat as YAML
      parsed = YAML.parse(raw);
    }
  } catch (err) {
    throw new VideoCutError(
      `配置文件解析失败: ${absPath}\n  ${err instanceof Error ? err.message : err}`,
    );
  }

  if (!parsed || typeof parsed !== 'object') {
    throw new VideoCutError(`配置文件格式错误: 应为对象，实际为 ${typeof parsed}`);
  }

  const config = parsed as Record<string, unknown>;

  if (!config.template || typeof config.template !== 'string') {
    throw new VideoCutError('配置文件缺少 "template" 字段（模板 ID）');
  }

  return {
    template: config.template,
    preset: typeof config.preset === 'string' ? config.preset : undefined,
    quality: typeof config.quality === 'string' ? config.quality : undefined,
    output:
      config.output && typeof config.output === 'object'
        ? (config.output as { filename?: string })
        : undefined,
    variables:
      config.variables && typeof config.variables === 'object'
        ? (config.variables as Record<string, unknown>)
        : {},
  };
}

/**
 * 获取项目配置文件所在目录（用于解析素材相对路径）
 */
export function getProjectDir(configPath: string): string {
  return path.dirname(path.resolve(configPath));
}

File: videocut/project/config_parser.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml

from videocut.errors import VideoCutError


@dataclass(slots=True)
class ProjectOutputConfig:
    filename: str | None = None


@dataclass(slots=True)
class ProjectConfig:
    template: str
    preset: str | None
    quality: str | None
    output: ProjectOutputConfig | None
    variables: dict[str, Any]


def parse_project_config(config_path: str | Path) -> ProjectConfig:
    abs_path = Path(config_path).resolve()
    if not abs_path.exists():
        raise VideoCutError(f"Config file does not exist: {abs_path}")

    raw = abs_path.read_text(encoding="utf-8")
    try:
        parsed = json.loads(raw) if abs_path.suffix.lower() == ".json" else yaml.safe_load(raw)
    except Exception as exc:  # noqa: BLE001
        raise VideoCutError(f"Failed to parse config: {abs_path}\n  {exc}") from exc

    if not isinstance(parsed, dict):
        raise VideoCutError(f"Config root must be an object, got {type(parsed).__name__}.")

    template = parsed.get("template")
    if not isinstance(template, str) or not template.strip():
        raise VideoCutError('Config must include a string "template" field.')

    output_raw = parsed.get("output")
    output = None
    if isinstance(output_raw, dict):
        output = ProjectOutputConfig(filename=output_raw.get("filename"))

    variables = parsed.get("variables")
    return ProjectConfig(
        template=template,
        preset=parsed.get("preset") if isinstance(parsed.get("preset"), str) else None,
        quality=parsed.get("quality") if isinstance(parsed.get("quality"), str) else None,
        output=output,
        variables=variables if isinstance(variables, dict) else {},
    )


def get_project_dir(config_path: str | Path) -> Path:
    return Path(config_path).resolve().parent

